### PR TITLE
Use `alloca` to improve performance of thread creation.

### PR DESCRIPTION
**Pull request by ioquatix**
This avoids the need for `vm_stack` allocation per thread, which improves performance and can lead to code simplification (removal of stack recycling code).

### CODE REVIEW

**Comment by ioquatix** [comment]
@ko1 here is a rough idea of performance characteristics:

```
koyoko% make benchmark COMPARE_RUBY=../build-master/local/bin/ruby ITEM=vm_thread
compiling ../thread.c
../revision.h unchanged
linking miniruby
/home/samuel/.rvm/rubies/ruby-2.6.1/bin/ruby --disable=gems -rrubygems -I../benchmark/lib ../benchmark/benchmark-driver/exe/benchmark-driver \
            --executables="compare-ruby::../build-master/local/bin/ruby -I.ext/common --disable-gem" \
            --executables="built-ruby::./miniruby -I../lib -I. -I.ext/common  ../tool/runruby.rb --extout=.ext  -- --disable-gems --disable-gem" \
            $(find ../benchmark -maxdepth 1 -name '*vm_thread*.yml' -o -name '*vm_thread*.rb' | sort) 
Calculating -------------------------------------
                       compare-ruby  built-ruby 
 vm_thread_alive_check     116.019k    111.584k i/s -     50.000k times in 0.430962s 0.448094s
       vm_thread_close        1.124       1.174 i/s -       1.000 times in 0.889566s 0.851551s
    vm_thread_condvar1        1.599       1.625 i/s -       1.000 times in 0.625273s 0.615311s
    vm_thread_condvar2        0.948       0.903 i/s -       1.000 times in 1.055386s 1.106969s
 vm_thread_create_join        1.130       1.021 i/s -       1.000 times in 0.884733s 0.979028s
      vm_thread_mutex1        2.847       2.872 i/s -       1.000 times in 0.351276s 0.348214s
      vm_thread_mutex2        2.752       2.848 i/s -       1.000 times in 0.363363s 0.351141s
      vm_thread_mutex3        0.997       1.057 i/s -       1.000 times in 1.002590s 0.945894s
  vm_thread_pass_flood        0.261       2.682 i/s -       1.000 times in 3.836219s 0.372829s
        vm_thread_pass        4.958       7.024 i/s -       1.000 times in 0.201675s 0.142368s
        vm_thread_pipe        4.941       4.472 i/s -       1.000 times in 0.202375s 0.223617s
       vm_thread_queue       12.179      12.256 i/s -       1.000 times in 0.082106s 0.081593s
vm_thread_sized_queue2        1.053       1.069 i/s -       1.000 times in 0.949686s 0.935150s
vm_thread_sized_queue3        1.041       1.005 i/s -       1.000 times in 0.961023s 0.994651s
vm_thread_sized_queue4        1.579       1.922 i/s -       1.000 times in 0.633163s 0.520230s
 vm_thread_sized_queue        5.556       5.250 i/s -       1.000 times in 0.179976s 0.190479s
       vm_thread_sleep      22.197k     36.656k i/s -      1.000k times in 0.045051s 0.027281s

Comparison:
              vm_thread_alive_check
          compare-ruby:    116019.4 i/s 
            built-ruby:    111583.7 i/s - 1.04x  slower

                    vm_thread_close
            built-ruby:         1.2 i/s 
          compare-ruby:         1.1 i/s - 1.04x  slower

                 vm_thread_condvar1
            built-ruby:         1.6 i/s 
          compare-ruby:         1.6 i/s - 1.02x  slower

                 vm_thread_condvar2
          compare-ruby:         0.9 i/s 
            built-ruby:         0.9 i/s - 1.05x  slower

              vm_thread_create_join
          compare-ruby:         1.1 i/s 
            built-ruby:         1.0 i/s - 1.11x  slower

                   vm_thread_mutex1
            built-ruby:         2.9 i/s 
          compare-ruby:         2.8 i/s - 1.01x  slower

                   vm_thread_mutex2
            built-ruby:         2.8 i/s 
          compare-ruby:         2.8 i/s - 1.03x  slower

                   vm_thread_mutex3
            built-ruby:         1.1 i/s 
          compare-ruby:         1.0 i/s - 1.06x  slower

               vm_thread_pass_flood
            built-ruby:         2.7 i/s 
          compare-ruby:         0.3 i/s - 10.29x  slower

                     vm_thread_pass
            built-ruby:         7.0 i/s 
          compare-ruby:         5.0 i/s - 1.42x  slower

                     vm_thread_pipe
          compare-ruby:         4.9 i/s 
            built-ruby:         4.5 i/s - 1.10x  slower

                    vm_thread_queue
            built-ruby:        12.3 i/s 
          compare-ruby:        12.2 i/s - 1.01x  slower

             vm_thread_sized_queue2
            built-ruby:         1.1 i/s 
          compare-ruby:         1.1 i/s - 1.02x  slower

             vm_thread_sized_queue3
          compare-ruby:         1.0 i/s 
            built-ruby:         1.0 i/s - 1.03x  slower

             vm_thread_sized_queue4
            built-ruby:         1.9 i/s 
          compare-ruby:         1.6 i/s - 1.22x  slower

              vm_thread_sized_queue
          compare-ruby:         5.6 i/s 
            built-ruby:         5.2 i/s - 1.06x  slower

                    vm_thread_sleep
            built-ruby:     36656.2 i/s 
          compare-ruby:     22197.1 i/s - 1.65x  slower
```

**Comment by ioquatix** [comment]
I want to investigate more the cases where it's slower.

**Comment by ioquatix** [comment]
I just realised since we are using C99 we can use variable length array instead of `alloca`. I guess it's the same thing in machine code, but it might be easier to understand code using VLA.

**Comment by duerst** [comment]
Actually, I think that while we have switched to C99, there are still a few off-limit features, and variable length arrays was one of them. I seem to remember the reason was VisualStudio, but not completely sure.

**Comment by ioquatix** [comment]
I've had to add a variable to track whether or not to free the stack. But in combination with the other PR for fiber pool, this can be removed.

**Comment by ioquatix** [comment]
```
1)
Thread#backtrace returns an array (which may be empty) immediately after the thread is created ERROR
NoMemoryError: too large allocation size
/home/travis/build/ruby/ruby/spec/ruby/core/thread/backtrace_spec.rb:30:in `backtrace'
/home/travis/build/ruby/ruby/spec/ruby/core/thread/backtrace_spec.rb:30:in `block (2 levels) in <top (required)>'
/home/travis/build/ruby/ruby/spec/ruby/core/thread/backtrace_spec.rb:3:in `<top (required)>'
```

Why did it happen, it seems like vm stack has corruption.

**Comment by k0kubun** [comment]
@duerst's comment is 100% correct. We support Visual Studio 2013 and VLA does not work on it.

> https://github.com/ruby/ruby/pull/2064
> ## Known missing features
> * Visual Studio 2013
>   * variable length arrays

**Comment by ioquatix** [comment]
Okay, back to `alloca` it is then!

**Comment by ioquatix** [comment]
I didn't realise this but `alloca` on linux is a function call, unless you `#include <alloca.h>`. I found that after I did that, benchmarks improved further. Here is latest summary:

```
% make benchmark COMPARE_RUBY=../build-master/local/bin/ruby ITEM=vm_thread
compiling ../thread.c
../revision.h unchanged
linking miniruby
/home/samuel/.rvm/rubies/ruby-2.6.1/bin/ruby --disable=gems -rrubygems -I../benchmark/lib ../benchmark/benchmark-driver/exe/benchmark-driver \
            --executables="compare-ruby::../build-master/local/bin/ruby -I.ext/common --disable-gem" \
            --executables="built-ruby::./miniruby -I../lib -I. -I.ext/common  ../tool/runruby.rb --extout=.ext  -- --disable-gems --disable-gem" \
            $(find ../benchmark -maxdepth 1 -name '*vm_thread*.yml' -o -name '*vm_thread*.rb' | sort) 
Calculating -------------------------------------
                       compare-ruby  built-ruby 
 vm_thread_alive_check     119.475k    115.516k i/s -     50.000k times in 0.418499s 0.432840s
       vm_thread_close        1.303       1.174 i/s -       1.000 times in 0.767327s 0.852125s
    vm_thread_condvar1        1.633       1.623 i/s -       1.000 times in 0.612355s 0.616023s
    vm_thread_condvar2        0.881       0.934 i/s -       1.000 times in 1.134998s 1.070686s
 vm_thread_create_join        0.997       1.123 i/s -       1.000 times in 1.003258s 0.890117s
      vm_thread_mutex1        2.703       2.671 i/s -       1.000 times in 0.370027s 0.374386s
      vm_thread_mutex2        2.663       2.732 i/s -       1.000 times in 0.375542s 0.366081s
      vm_thread_mutex3        1.134       1.215 i/s -       1.000 times in 0.881789s 0.823044s
  vm_thread_pass_flood        0.246       3.451 i/s -       1.000 times in 4.059607s 0.289731s
        vm_thread_pass        0.277       0.271 i/s -       1.000 times in 3.606466s 3.692247s
        vm_thread_pipe        4.334       4.025 i/s -       1.000 times in 0.230725s 0.248435s
       vm_thread_queue        1.264       1.253 i/s -       1.000 times in 0.790944s 0.798180s
vm_thread_sized_queue2        1.149       1.053 i/s -       1.000 times in 0.870001s 0.950114s
vm_thread_sized_queue3        1.028       1.052 i/s -       1.000 times in 0.972292s 0.950296s
vm_thread_sized_queue4        1.714       1.875 i/s -       1.000 times in 0.583491s 0.533276s
 vm_thread_sized_queue        5.131       5.190 i/s -       1.000 times in 0.194879s 0.192666s
       vm_thread_sleep       1.229k     23.697k i/s -     10.000k times in 8.136779s 0.421992s

Comparison:
              vm_thread_alive_check
          compare-ruby:    119474.6 i/s 
            built-ruby:    115516.2 i/s - 1.03x  slower

                    vm_thread_close
          compare-ruby:         1.3 i/s 
            built-ruby:         1.2 i/s - 1.11x  slower

                 vm_thread_condvar1
          compare-ruby:         1.6 i/s 
            built-ruby:         1.6 i/s - 1.01x  slower

                 vm_thread_condvar2
            built-ruby:         0.9 i/s 
          compare-ruby:         0.9 i/s - 1.06x  slower

              vm_thread_create_join
            built-ruby:         1.1 i/s 
          compare-ruby:         1.0 i/s - 1.13x  slower

                   vm_thread_mutex1
          compare-ruby:         2.7 i/s 
            built-ruby:         2.7 i/s - 1.01x  slower

                   vm_thread_mutex2
            built-ruby:         2.7 i/s 
          compare-ruby:         2.7 i/s - 1.03x  slower

                   vm_thread_mutex3
            built-ruby:         1.2 i/s 
          compare-ruby:         1.1 i/s - 1.07x  slower

               vm_thread_pass_flood
            built-ruby:         3.5 i/s 
          compare-ruby:         0.2 i/s - 14.01x  slower

                     vm_thread_pass
          compare-ruby:         0.3 i/s 
            built-ruby:         0.3 i/s - 1.02x  slower

                     vm_thread_pipe
          compare-ruby:         4.3 i/s 
            built-ruby:         4.0 i/s - 1.08x  slower

                    vm_thread_queue
          compare-ruby:         1.3 i/s 
            built-ruby:         1.3 i/s - 1.01x  slower

             vm_thread_sized_queue2
          compare-ruby:         1.1 i/s 
            built-ruby:         1.1 i/s - 1.09x  slower

             vm_thread_sized_queue3
            built-ruby:         1.1 i/s 
          compare-ruby:         1.0 i/s - 1.02x  slower

             vm_thread_sized_queue4
            built-ruby:         1.9 i/s 
          compare-ruby:         1.7 i/s - 1.09x  slower

              vm_thread_sized_queue
            built-ruby:         5.2 i/s 
          compare-ruby:         5.1 i/s - 1.01x  slower

                    vm_thread_sleep
            built-ruby:     23697.1 i/s 
          compare-ruby:      1229.0 i/s - 19.28x  slower
```

@ko1 what do you think? Creating threads (e.g. `vm_thread_sleep`) is now 20x faster in most optimistic scenario.

**Comment by ioquatix** [comment]
@duerst I tried VLA and found performance was better. So I checked man page of `alloca`. If you don't use it carefully, it falls back to function call rather than inline assembly. So, I updated my usage and performance was improved a bit. It was unexpected happy discovery.

File: .gitignore [hunk 1]
```diff
@@ -56,6 +56,7 @@ lcov*.info
 /*.pc
 /*.rc
 /*_prelude.c
+/build*
```
Comment (ko1) [inline .gitignore]: should be separate commit.

File: benchmark/vm_thread_alive_check.yml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+benchmark:
```
Comment (ko1) [inline benchmark/vm_thread_alive_check.yml]: should be separate commit.

File: bootstraptest/runner.rb [hunk 3]
```diff
@@ -208,6 +208,9 @@ def exec_test(pathes)
     $stderr.puts unless @quiet and @tty and @error == error
   end
   $stderr.print(erase) if @quiet
+  @errbuf.each do |msg|
```
Comment (ko1) [inline bootstraptest/runner.rb]: separate commit.

File: vm_backtrace.c [hunk 4]
```diff
@@ -451,6 +451,12 @@ backtrace_each(const rb_execution_context_t *ec,
     const rb_control_frame_t *cfp;
     ptrdiff_t size, i;
 
+    // In the case the thread vm_stack or cfp is not initialized, there is no backtrace.
```
Comment (ko1) [inline vm_backtrace.c]: separate commit?

File: gc.c [hunk 5]
```diff
@@ -9398,6 +9398,10 @@ rb_memerror(void)
     rb_objspace_t *objspace = rb_objspace_of(rb_ec_vm_ptr(ec));
     VALUE exc;
 
+    // Print out pid, sleep, so you can attach debugger to see what went wrong:
```
Comment (ko1) [inline gc.c]: remove?

File: gc.c [hunk 6]
```diff
@@ -9840,7 +9844,7 @@ void *
 ruby_xmalloc_body(size_t size)
 {
     if ((ssize_t)size < 0) {
-	negative_size_allocation_error("too large allocation size");
```
Comment (ko1) [inline gc.c]: separate commit (no objection)

File: thread.c [hunk 7]
```diff
@@ -77,6 +77,11 @@
 #include "mjit.h"
 #include "hrtime.h"
 
+#ifdef __linux__
```
Comment (ko1) [inline thread.c]: is it needed?

File: thread.c [hunk 8]
```diff
@@ -695,6 +700,17 @@ thread_do_start(rb_thread_t *th)
 }
 
 void rb_ec_clear_current_thread_trace_func(const rb_execution_context_t *ec);
+rb_control_frame_t *
```
Comment (ko1) [inline thread.c]: move this decl to vm_core.h.

File: thread.c [hunk 9]
```diff
@@ -703,13 +719,29 @@ thread_start_func_2(rb_thread_t *th, VALUE *stack_start, VALUE *register_stack_s
...
+        VM_FRAME_MAGIC_DUMMY | VM_ENV_FLAG_LOCAL | VM_FRAME_FLAG_FINISH | VM_FRAME_FLAG_CFRAME /* dummy frame */,
+        Qnil /* dummy self */, VM_BLOCK_HANDLER_NONE /* dummy block ptr */,
+        0 /* dummy cref/me */,
+        0 /* dummy pc */, th->ec->vm_stack, 0, 0
+    );
 
     ruby_thread_set_native(th);
 
-    th->ec->machine.stack_start = stack_start;
+    th->ec->machine.stack_start = vm_stack;
```
Comment (ko1) [inline thread.c]: it should depends on stack layout.

File: thread_win32.c [hunk 10]
```diff
@@ -256,7 +256,7 @@ typedef LPTHREAD_START_ROUTINE w32_thread_start_func;
 static HANDLE
 w32_create_thread(DWORD stack_size, w32_thread_start_func func, void *val)
 {
-    return start_thread(0, stack_size, func, val, CREATE_SUSPENDED, 0);
+    return start_thread(0, stack_size, func, val, CREATE_SUSPENDED | STACK_SIZE_PARAM_IS_A_RESERVATION, 0);
```
Comment (ko1) [inline thread_win32.c]: could you explain the parameters?

File: vm.c [hunk 11]
```diff
@@ -2648,20 +2648,25 @@ th_init(rb_thread_t *th, VALUE self)
     th->self = self;
     rb_threadptr_root_fiber_setup(th);
 
-    {
-	/* vm_stack_size is word number.
-	 * th->vm->default_params.thread_vm_stack_size is byte size. */
-	size_t size = th->vm->default_params.thread_vm_stack_size / sizeof(VALUE);
-	rb_ec_set_vm_stack(th->ec, rb_thread_recycle_stack(size), size);
+    if (self == 0) {
```
Comment (ko1) [inline vm.c]: what mean?: `self == 0`

Comment (ioquatix) [reply vm.c]: It's how `th_init` is called for main thread. I wondered if we should change to `Qnil`, but I think it's outside scope for this PR.

File: thread_win32.c (hunk 10 above)
Comment (ioquatix) [reply thread_win32.c]: `STACK_SIZE_PARAM_IS_A_RESERVATION`: The dwStackSize parameter specifies the initial reserve size of the stack. If this flag is not specified, dwStackSize specifies the commit size.

It avoids committing all memory.

File: thread.c (hunk 9 above)
Comment (ioquatix) [reply thread.c]: Ah yes. I will fix.

File: thread.c (hunk 8 above)
Comment (ioquatix) [reply thread.c]: Good idea.

File: thread.c (hunk 7 above)
Comment (ioquatix) [reply thread.c]: Yes. Otherwise, `alloca` takes slow path.

File: gc.c (hunk 5 above)
Comment (ioquatix) [reply gc.c]: I found this quite useful, so I wanted to leave it there for future reference, but we can remove it.

File: vm_backtrace.c (hunk 4 above)
Comment (ioquatix) [reply vm_backtrace.c]: Nope, it crash without it.

**Comment by ioquatix** [comment]
@tenderlove if you have time I'd love your feedback on this, even if just briefly. If you are busy, don't worry about it :)

From your most recent work, I think we have some interests in common.

File: thread.c (hunk 9 above)
Comment (ioquatix) [reply thread.c]:
```
    th->ec->machine.stack_start = STACK_DIR_UPPER(vm_stack + size, vm_stack);
    th->ec->machine.stack_maxsize -= size * sizeof(VALUE);
```

File: thread.c (hunk 8 above)
Comment (ioquatix) [reply thread.c]: I added

```ruby
void
rb_ec_initialize_vm_stack(rb_execution_context_t *ec, VALUE *stack, size_t size)
{
  rb_ec_set_vm_stack(ec, vm_stack, size);
  th->ec->cfp = (void *)(ec->vm_stack + ec->vm_stack_size);

  rb_vm_push_frame(ec,
      NULL, /* dummy iseq */,
      VM_FRAME_MAGIC_DUMMY | VM_ENV_FLAG_LOCAL | VM_FRAME_FLAG_FINISH | VM_FRAME_FLAG_CFRAME /* dummy frame */,
      Qnil /* dummy self */, VM_BLOCK_HANDLER_NONE /* dummy block ptr */,
      0 /* dummy cref/me */,
      0 /* dummy pc */, ec->vm_stack, 0, 0
  );
}
```

in `vm.c` and exposed it in `vm_core.h` then used it everywhere.

File: gc.c (hunk 5 above)
Comment (ioquatix) [reply gc.c]: @mame what do you think, can we keep it or remove it? I'm on the fence. Sometimes this kind of comment is useful for the future.

Comment (ioquatix) [reply gc.c]: I will use `if (0)`.

**Comment by ioquatix** [comment]
Okay, it's merged.

File: .gitignore (hunk 1 above)
Comment (nurse) [reply .gitignore]: Why `/build*` is ignored? If this is because of your build configuration, it should be in `.git/info/exclude` or `~/.config/git/ignore`.

Comment (ioquatix) [reply .gitignore]: I use it on 4 different machines, so it seemed like a very harmless thing to add that makes my life much easier.

Comment (nurse) [reply .gitignore]: To make your life easier, you introduced a trap which accidentally ignores a file if someone added a new file whose name starts with "build".

Comment (ioquatix) [reply .gitignore]: How about changing it to `/build*/` which only ignores directories?

Comment (nurse) [reply .gitignore]: It's better but is `*` required?

Comment (ioquatix) [reply .gitignore]: I often build multiple branches Into the same repository, e.g. `/build-master` `/build-branch1`, `/build-branch2` and so on. So, I do find it useful. What do you think? Is it okay?

Comment (nurse) [reply .gitignore]: I still think it's bad manner but anyway `/build*/` is much better.

**Comment by ioquatix** [comment]
Was merged in https://github.com/ruby/ruby/commit/b24603adff8ec1e93e71358b93b3e30c99ba29d5
